---
title: "SummarizedExperiment"
author: "[Michael Love](http://mikelove.github.io)"
output: html_document
---

```{r message=FALSE}
library(SummarizedExperiment)
colData <- data.frame(sample=factor(1:6),
                      condition=factor(c("A","A","B","B","C","C")),
                      treated=factor(rep(0:1,3)))
colData
```

```{r}
library(EnsDb.Hsapiens.v86)
txdb <- EnsDb.Hsapiens.v86
g <- genes(txdb)
g <- keepStandardChromosomes(g, pruning.mode="coarse")
rowRanges <- g[1:10]
```

```{r}
exprs <- matrix(rnorm(6 * 10), ncol=6, nrow=10)
se <- SummarizedExperiment(exprs, colData=colData, rowRanges=rowRanges)
se
```

```{r}
query <- GRanges("1", IRanges(25000,40000))
se.sub <- se[overlapsAny(se, query),]
```

```{r}
rowRanges(se.sub)
assay(se.sub)
```

```{r}
seqinfo(se)
```

We previously introduced the computational project, called 
[recount2](https://jhubiostatistics.shinyapps.io/recount/), which performs a
basic summarization of public data sets with gene expression data. We
will use data from *recount2* again.

```{r}
url <- "http://duffel.rail.bio/recount/SRP046226/rse_gene.Rdata"
file <- "asthma.rda"
if (!file.exists(file)) download.file(url, file)
load(file)
ls()
```

```{r}
source("my_scale_counts.R")
rse <- my_scale_counts(rse_gene)
```

```{r}
colData(rse)[,1:6]
rse$characteristics[[1]]
rse$condition <- sapply(rse$characteristics, `[`, 3)
rse$treatment <- sapply(rse$characteristics, `[`, 4)
```

```{r}
rowRanges(rse)
```

```{r}
library(magrittr)
rse$condition %<>% (function(x) {
  factor(sub("-",".", sub("disease state: (.*)","\\1",x) ))
  })
rse$treatment %<>% (function(x) factor(sub("treatment: (.*)","\\1",x)))
```

```{r}
rse$condition
rse$treatment
```

```{r}
library(DESeq2)
dds <- DESeqDataSet(rse, ~condition + treatment)
```

```{r}
vsd <- vst(dds, blind=FALSE)
```

```{r}
library(matrixStats)
rv <- rowVars(assay(vsd))
```

```{r}
anno.col <- as.data.frame(colData(vsd)[,c("condition","treatment")])
anno.col
```

```{r}
library(pheatmap)
pheatmap(assay(vsd)[head(order(rv, decreasing=TRUE),100),],
         annotation_col=anno.col,
         show_rownames=FALSE, show_colnames=FALSE)
```
